File: kotlin-react-router/karakum-legacy/remix-run-router/karakum/plugins/convertOverriddenPropertySignature.js
import ts from "typescript";

export default function (node, context, render) {
    if (
        ts.isPropertySignature(node)
        && node?.parent?.parent?.parent // PropertySignature / TypeLiteral / IntersectionType / TypeAliasDeclaration
        && ts.isTypeAliasDeclaration(node.parent.parent.parent)
        && ts.isIdentifier(node.parent.parent.parent.name)
        && (
            node.parent.parent.parent.name.text === "AgnosticDataNonIndexRouteObject"
            || node.parent.parent.parent.name.text === "AgnosticDataIndexRouteObject"
        )
        && ts.isIdentifier(node.name)
        && (
            node.name.text === "id"
            || node.name.text === "children"
        )
    ) {
        const readonly = node.modifiers?.find(modifier => modifier.kind === ts.SyntaxKind.ReadonlyKeyword)

        const modifier = readonly
            ? "val "
            : "var "

        const name = render(node.name)

        let type = render(node.type)

        const nullable = node.questionToken
            || node.name.text === "id" // id property is nullable and var in parent type

        return `override ${modifier}${name}: ${type}${nullable ? "?" : ""}`
    }
    return null
}

File: kotlin-electron/karakum/src/jsMain/kotlin/electron/karakum/plugins/convertOverriddenPropertySignature.kt
package electron.karakum.plugins

import electron.karakum.plugins.SignatureOverrideConfig.Alias
import electron.karakum.plugins.SignatureOverrideConfig.Overlap
import electron.karakum.plugins.SignatureOverrideConfig.Preserve
import electron.karakum.util.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.extension.renderNullable
import io.github.sgrishchenko.karakum.util.getParentOrNull
import typescript.SyntaxKind
import typescript.asArray
import typescript.isIdentifier
import typescript.isInterfaceDeclaration
import typescript.isPropertySignature

private sealed interface SignatureOverrideConfig {
    object Preserve : SignatureOverrideConfig
    class Alias(val alias: String) : SignatureOverrideConfig
    class Overlap(val alias: String, val parentType: String) : SignatureOverrideConfig
}

private val overriddenProps = mapOf<String, Map<String, SignatureOverrideConfig>>(
    "KeyboardInputEvent" to mapOf(
        "type" to Overlap("keyboardType", "InputEventType"),
    ),
    "MouseInputEvent" to mapOf(
        "type" to Overlap("mouseType", "InputEventType"),
    ),
    "MouseWheelInputEvent" to mapOf(
        "type" to Overlap("mouseWheelType", "InputEventType"),
    ),
)

val convertOverriddenPropertySignature = createPlugin { node, context, render ->
    nullable {
        ensure(isPropertySignature(node))

        val interfaceNode = ensureNotNull(node.getParentOrNull())
        ensure(isInterfaceDeclaration(interfaceNode))

        val interfaceName = interfaceNode.name
        val interfaceOverrides = ensureNotNull(overriddenProps[interfaceName.text])

        val propertyName = node.name
        ensure(isIdentifier(propertyName))
        val propertyOverride = ensureNotNull(interfaceOverrides[propertyName.text])

        val readonly = node.modifiers?.asArray()?.find { it.kind == SyntaxKind.ReadonlyKeyword }

        val modifier = if (readonly != null) "val " else "var "

        val name = render(node.name)

        val isOptional = node.questionToken != null

        val type = renderNullable(node.type, isOptional, context, render)

        when (propertyOverride) {
            is Overlap -> {
                val alias = propertyOverride.alias
                val jsName = "@JsName(\"$name\")"
                val aliasSignature = "${modifier}${alias}: $type"
                val deprecation = "@Deprecated(message = \"use ${alias}\", level = DeprecationLevel.HIDDEN)"
                val signature = "override ${modifier}${name}: ${propertyOverride.parentType}"

                arrayOf(
                    deprecation,
                    signature,
                    jsName,
                    aliasSignature,
                ).joinToString("\n")
            }
            is Alias -> {
                val alias = propertyOverride.alias
                val jsName = "@JsName(\"$name\")"
                val aliasSignature = "${modifier}${alias}: $type"

                arrayOf(
                    jsName,
                    aliasSignature,
                ).joinToString("\n")
            }
            Preserve -> "override ${modifier}${name}: $type"
        }
    }
}
